### ISSUE TITLE
Work: Split overloaded src/unilab/utils ownership

### ISSUE THREAD

**Issue by TATP-233** [Motphys/UniLab#334]
### Work type
infra

### Area
other

### Problem
`src/unilab/utils` currently carries too many unrelated responsibilities and has become the default place for new helper files. This makes ownership unclear, increases coupling across env / backend / runner / algo layers, and encourages future code to bypass the intended owner modules by adding more files under `utils`.

Current files under `src/unilab/utils` span multiple owner areas:

| Current file | Observed responsibility | Proposed owner |
| --- | --- | --- |
| `algo_utils.py` | registry bootstrap and actor construction | `src/unilab/training/registry.py` for registry bootstrap; `src/unilab/algos/torch/common/factory.py` or existing algo common owner for actor construction |
| `device_utils.py` | device selection and env dim derivation for torch runners | `src/unilab/algos/torch/common/device.py` or runner-owned helper close to torch algorithms |
| `experiment_tracking.py` | W&B / experiment metadata / tracking lifecycle | `src/unilab/training/tracking.py` or `src/unilab/training/experiment_tracking.py` |
| `logging_common.py` | shared rich / W&B logger primitives | `src/unilab/training/logging/common.py` |
| `offpolicy_logger.py` | off-policy training logger | `src/unilab/algos/torch/offpolicy/logging.py`, with shared pieces imported from `src/unilab/training/logging` |
| `onpolicy_logger.py` | on-policy training logger | `src/unilab/algos/torch/common/onpolicy_logging.py` or `src/unilab/training/logging/onpolicy.py` |
| `rsl_rl_compat.py` | rsl-rl version/config compatibility | `src/unilab/algos/torch/rsl_rl_compat.py` |
| `rsl_rl_vec_env_wrapper.py` | rsl-rl vector env wrapper and obs flattening bridge | `src/unilab/algos/torch/rsl_rl_vec_env_wrapper.py` |
| `obs_utils.py` | obs dict flatten/split/dim contract helpers | `src/unilab/base/obs.py` or `src/unilab/training/obs.py`; must remain contract-oriented and backend-neutral |
| `final_observation.py` | terminal observation contract resolution | `src/unilab/base/final_observation.py` or `src/unilab/training/final_observation.py` |
| `reward_utils.py` | reward config extraction / resolution | `src/unilab/config/rewards.py` or `src/unilab/training/rewards.py` |
| `xml_utils.py` | XML scene mutation, named body ids, visual materialization | split between `src/unilab/base/backend/xml.py` and backend-specific modules under `src/unilab/base/backend/`; keep asset/XML access on cold paths only |
| `render_many.py` | MuJoCo rendering / frame generation | `src/unilab/base/backend/rendering.py` or `src/unilab/base/backend/mujoco_rendering.py` |
| `viser_scene.py` | MuJoCo + viser visualization scene | `src/unilab/base/backend/mujoco_viser.py` or `src/unilab/visualization/viser_scene.py` if a visualization package is introduced |
| `math_utils.py` | numpy quaternion/math helpers used by envs and scripts | keep only if declared as cross-cutting numeric utilities, or move env-specific quaternion helpers to `src/unilab/envs/common/math.py` |
| `torch_utils.py` | tensor/array conversion helpers | keep only if declared as cross-cutting tensor utilities, or move to `src/unilab/algos/torch/common/tensor.py` |
| `hardware_monitor.py` | torch hardware metrics | `src/unilab/training/hardware_monitor.py` or logger/tracking owner |
| `run_utils.py` | process/runtime env helpers | `src/unilab/training/run.py` or script-owned helper if only scripts use it |

### Deliverable
- Publish an owner map for every current `src/unilab/utils/*.py` file.
- Define an explicit rule: `src/unilab/utils` may only contain small, dependency-light, backend-neutral helpers with no owner-specific lifecycle or contract semantics.
- Move owner-specific utilities into the owning package listed above, updating imports intentionally.
- Add temporary compatibility shims only where needed to avoid a risky single large migration; each shim must include a removal target or follow-up issue.
- Update tests and docs that reference old `unilab.utils.*` paths.
- Add a short developer note documenting where future helpers should live.

### Proposed migration plan
1. **Inventory and guardrail**
   - Generate an import inventory with `rg "unilab\.utils|from unilab.utils|import unilab.utils"`.
   - Add a small developer doc section that states the allowed scope of `src/unilab/utils`.
   - Decide whether `math_utils.py` and `torch_utils.py` stay as the only cross-cutting utilities or move under env / torch owners.

2. **Low-risk training/logging moves**
   - Move `experiment_tracking.py`, `logging_common.py`, `offpolicy_logger.py`, `onpolicy_logger.py`, `hardware_monitor.py`, and `run_utils.py` under training / algo logging owners.
   - Update scripts and algorithm runners to import from the new owner paths.
   - Keep shims in `unilab.utils` during the transition if external scripts are likely to import these names.

3. **Algorithm owner moves**
   - Move `rsl_rl_compat.py` and `rsl_rl_vec_env_wrapper.py` under the torch rsl-rl owner.
   - Split `algo_utils.py` so registry bootstrap moves to training/common registry ownership, while actor construction moves to torch algo common ownership.
   - Update `scripts/train_rsl_rl.py`, `scripts/play_viser.py`, APPO/off-policy runners, and affected tests.

4. **Contract/config owner moves**
   - Move `obs_utils.py` and `final_observation.py` to the env/training contract owner selected in step 1.
   - Move `reward_utils.py` to config/training reward owner.
   - Validate `NpEnvState.obs` dict handling, `obs_groups_spec` dimensions, reward Hydra injection, and terminal observation behavior near the changed imports.

5. **Backend/rendering/XML moves**
   - Split `xml_utils.py` into backend-neutral XML helpers and backend-specific MuJoCo/Motrix mutation helpers.
   - Move `render_many.py` and `viser_scene.py` to backend/rendering or visualization ownership.
   - Ensure XML / asset / model metadata access remains on init/materialization/cache paths only, never in `step`, `reset`, domain randomization, or hot runner loops.

6. **Cleanup and enforcement**
   - Remove compatibility shims once internal imports are migrated.
   - Keep `src/unilab/utils/__init__.py` minimal or empty; avoid re-exporting owner-specific APIs.
   - Add or update tests that fail if new owner-specific imports are reintroduced under `src/unilab/utils`.

### Definition of done
- `src/unilab/utils` contains only explicitly approved cross-cutting helpers, or is empty except for compatibility shims with removal tracking.
- Every moved file has a clear owner package and import path.
- Owner-specific utilities no longer live in the shared utils package.
- Training scripts only assemble flows and do not accumulate backend / env / reward business rules during the migration.
- Backend-specific logic remains inside backend / env adaptation layers.
- Reward config handling remains Hydra/config driven.
- No hot-path asset/XML/model metadata access is introduced.
- Tests and docs no longer point users at `unilab.utils` as the default extension point.

### Dependencies and blockers
None known. Coordinate with active PRs that import from `src/unilab/utils` before moving modules, because current references exist across `scripts/`, `src/unilab/algos`, `src/unilab/base/backend`, `src/unilab/training`, `tests/`, and docs.

### Proposed owner
TBD

### Validation plan
- Run `rg "unilab\.utils|from unilab.utils|import unilab.utils" src scripts tests docs` before and after each migration stage.
- Run targeted tests around the changed owner boundary, for example:
  - `uv run pytest tests/utils/test_obs_utils.py tests/utils/test_final_observation.py`
  - `uv run pytest tests/config/test_reward_injection.py tests/base/test_reward_override.py`
  - `uv run pytest tests/algos/test_rsl_rl_runner.py tests/scripts/test_train_scripts.py`
  - `uv run pytest tests/utils/test_render_many.py tests/utils/test_viser_scene.py tests/base/test_sim_backend.py`
- Run `uv run ruff check` and the repository CI-equivalent command before closing.
- Confirm no new import from `src/unilab/utils` is introduced for owner-specific code.

**Comment by TATP-233** [comment]
https://github.com/unilabsim/UniLab/pull/342
